vpr/src/device/rr_graph_obj.h

```diff
@@ -0,0 +1,468 @@
+/************************************************************************
+ * This file introduces a class to model a Routing Resource Graph (RRGraph or RRG) 
+ * which is widely used by placers, routers, analyzers etc.
+ * A Routing Resource Graph (RRGraph or RRG) consists of a number of nodes and edges.
+ *
+ * Each node represents a routing resource, which could be 
+ * 1. a routing track in X-direction or Y-direction (CHANX or CHANY) 
+ * 2. an input or an output of a logic block (IPIN or OPIN) 
+ * 3. a virtual source or sink node (SOURCE or SINK), which are starting/ending points of routing trees.
+ *
+ * Each edge represents a switch between routing resources, which could be
+ * 1. a multiplexer
+ * 2. a tri-state buffer
+ * The switch information are categorized in the rr_switch_inf of RRGraph class.
+ *
+ * IMPORTANT:
+ *   The RRGraph is designed to be a read-only database/graph, once created.
+ *   Placement and routing should not change any attributes of RRGraph.
+ *   Any placement and routing results should be stored in other data structures, such as PlaceContext and RoutingContext. 
+ *
+ * How to use the RRGraph data structure 
+ * =====================================
+ * To iterate over the nodes/edges/switches/segments in a RRGraph, 
+ *    using a range-based loop is suggested.
+ *  -----------------------------------------------------------------
+ *    Example: iterate over all the nodes
+ *      for (RRNodeId node : nodes()) {
+ *        // Do something with each node
+ *      }
+ *
+ *      for (RREdgeId edge : edges()) {
+ *        // Do something with each edge
+ *      }
+ *
+ *      for (RRSwitchId switch : switches()) {
+ *        // Do something with each switch
+ *      }
+ *
+ *      for (RRSegmentId segment : segments()) {
+ *        // Do something with each segment
+ *      }
+ *
+ * Access to a node/edge/switch/segment, please use the StrongId created
+ *  -----------------------------------------------------------------
+ *    For node, use RRNodeId
+ *    For edge, use RREdgeId
+ *    For switch, use RRSwitchId
+ *    For segment, use RRSegmentId
+ *
+ *    These are the unique identifier for each data type.
+ *    To check if your id is valid or not, use the INVALID() function of StrongId class.
+ *    Example:
+ *       if (node_id == RRNodeId::INVALID()) {
+ *       }  
+ *
+ *  
+ * Detailed description on the RRGraph data structure
+ * ==================================================
+ *
+ * Node-related data 
```

Comment:
These are good comments describing the different bits of information stored in the RR graph.

However I think they need some changes. Keep in mind that there are multiple audiences of people who will be reading them:

1. Developers trying to use query the RR graph for information
This is by far the most common case, so we want to have most of the comments targeting this audience. This is where the file-level comments should outline what an RR graph is, how it's logically represented (e.g. what is an RR nodes's location?) and give some tutorial content on how to use the RR graph class (e.g. how to loop through nodes and associated edges).
There should also be comments on each of the public accessor member functions outlining what they represent/return, arguments etc. These per-member function comments serve as the detailed reference of how to use the class.


2. Developers trying to build/create an RR graph
This is a much smaller subset of developers, who will be creating the RR graphs (e.g. modifying the RR graph builder/loader).
You will want comments on the pubic mutator member functions outlining what they do, and likely some tutorial content covering how you would go about modifying/building an RR graph.

3. Developers changing the RR graph class implementation
This is an extremely small group of developers, but its important to document any key invariants/assumptions and motivations behind various design decisions. These should likely be mostly in the private data member section. Its also good to emphasize what **not** to do (e.g. put router state in the RR graph).

(Note that this is also why the various sections of the class are ordered as they are, first public accessors, then public mutators, then private implementation details. Most people only need the public accessor interface and can stop reading after that.)

As they currently stand, the comments which follow reference the private data member names. But this is only relevant to group (3) and not the much larger groups (1) and (2). So I think much of this information should be shifted to per-member function comments on the public interface.
Any comments which address the implementation details (e.g. node_lookup_) should be near the actual private data member definition.